peers/app_request_network.go

```diff
@@ -103,83 +100,99 @@ func NewNetwork(
-	// gossip will enable connecting to all the remaining nodes in the network.
-	var beaconIPs, beaconIDs []string
-
-	peers, err := infoClient.Peers(context.Background())
-	if err != nil {
-		logger.Error(
-			"Failed to get peers",
-			zap.Error(err),
-		)
-		return nil, nil, err
+	arNetwork := &AppRequestNetwork{
+		Network:    testNetwork,
+		Handler:    handler,
+		infoClient: infoClient,
+		logger:     logger,
+		lock:       new(sync.Mutex),
 	}
 
-	// Randomly select peers to connect to until we have numInitialTestPeers
-	indices := rand.Perm(len(peers))
-	for _, index := range indices {
-		// Do not attempt to connect to private peers
-		if len(peers[index].PublicIP) == 0 {
-			continue
-		}
-		beaconIPs = append(beaconIPs, peers[index].PublicIP)
-		beaconIDs = append(beaconIDs, peers[index].ID.String())
-		if len(beaconIDs) == numInitialTestPeers {
-			break
+	// Manually connect to the validators of each of the source subnets.
+	// We return an error if we are unable to connect to sufficient stake on any of the subnets.
+	// Sufficient stake is determined by the Warp quora of the configured supported destinations,
+	// or if the subnet supports all destinations, by the quora of all configured destinations.
+	for _, sourceBlockchain := range cfg.SourceBlockchains {
+		// Get the list of destination chains that this source may relay to
+		var destinationBlockchainIDs []ids.ID
+		if supportedDsts := sourceBlockchain.GetSupportedDestinations(); supportedDsts.Len() > 0 {
+			destinationBlockchainIDs = supportedDsts.List()
+		} else {
+			for _, dst := range cfg.DestinationBlockchains {
+				destinationBlockchainIDs = append(destinationBlockchainIDs, dst.GetBlockchainID())
+			}
 		}
-	}
-	if len(beaconIPs) == 0 {
-		logger.Error(
-			"Failed to find any peers to connect to",
-			zap.Error(err),
-		)
-		return nil, nil, err
-	}
-	if len(beaconIPs) < numInitialTestPeers {
-		logger.Warn(
-			"Failed to find a full set of peers to connect to on startup",
-			zap.Int("connectedPeers", len(beaconIPs)),
-			zap.Int("expectedConnectedPeers", numInitialTestPeers),
-		)
-	}
 
-	for i, beaconIDStr := range beaconIDs {
-		beaconID, err := ids.NodeIDFromString(beaconIDStr)
-		if err != nil {
-			logger.Error(
-				"Failed to parse beaconID",
-				zap.String("beaconID", beaconIDStr),
-				zap.Error(err),
+		var connectedValidators *ConnectedCanonicalValidators
+		if sourceBlockchain.GetBlockchainID() == constants.PrimaryNetworkID {
+			// For the primary network, connect to the validators of the destination chains
+			for _, destinationBlockchainID := range destinationBlockchainIDs {
+				subnetID, err := pChainClient.ValidatedBy(context.Background(), destinationBlockchainID)
+				if err != nil {
+					logger.Error(
+						"Failed to get subnet ID",
+						zap.Error(err),
+					)
+					return nil, nil, err
+				}
+				connectedValidators, err = arNetwork.ConnectToCanonicalValidators(
+					validators.NewCanonicalValidatorClient(logger, pChainClient),
```

Comment:
Does the `CanonicalValidatorClient` need to be newly created for each loop? Seems like we can move this out of the for loop

Reply:
or actually we can save `pChainClient` in `AppRequestNetwork` struct and don't need to pass that value in for `ConnectToCanonicalValidators`

Reply:
Good calls. Fixed.